org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/BuildManager.xtend

```diff
@@ -75,20 +81,40 @@ class BuildManager {
             project2deleted.put(projectManager, deleted)
         }
         val sortedDescriptions = sortByDependencies(project2dirty.keySet + project2deleted.keySet)
-        val result = newArrayList()
         for(ProjectDescription it: sortedDescriptions) {
             val projectManager = workspaceManager.getProjectManager(name)
             val projectDirty = project2dirty.get(it).toList
             val projectDeleted = project2deleted.get(it).toList
-            val partialResult = projectManager.doBuild(projectDirty, projectDeleted, result, cancelIndicator)
+            val partialResult = projectManager.doBuild(projectDirty, projectDeleted, unreportedDeltas, cancelIndicator)
             allDirty.addAll(partialResult.affectedResources.map[uri])
             this.dirtyFiles -= projectDirty
             this.deletedFiles -= projectDeleted
-            result.addAll(partialResult.affectedResources)
+            // prior builds could have been canceled, so their result has not been returned,
+            // but the projectManager already has updated its state
+            mergeWithUnreportedDeltas(partialResult.affectedResources)
         }
+        val result = unreportedDeltas
+		unreportedDeltas = newArrayList
         return result
     }
     
+    protected def mergeWithUnreportedDeltas(List<IResourceDescription.Delta> newDeltas) {
```

Comment:
Please add `@since` tag. New protected methods in public API are considered new API.

Reply:
Done